helix-term/src/ui/statusline.rs

```diff
@@ -0,0 +1,312 @@
+            .for_each(|render| render(context, write_center));
+
+        // Width of the empty space between the left and center area and between the center and right area.
+        let spacing = 1u16;
+
+        let edge_width = context.parts.left.width().max(context.parts.right.width()) as u16;
+        let center_max_width = viewport.width.saturating_sub(2 * edge_width + 2 * spacing);
+        let center_width = center_max_width.min(context.parts.center.width() as u16);
+
+        surface.set_spans(
+            viewport.x + viewport.width / 2 - center_width / 2,
+            viewport.y,
+            &context.parts.center,
+            center_width,
+        );
+    }
+
+    fn append(buffer: &mut Spans, text: String, base_style: &Style, style: Option<Style>) {
+        buffer.0.push(Span::styled(
+            text,
+            style.map_or(*base_style, |s| (*base_style).patch(s)),
+        ));
+    }
+
+    fn get_render_function<F>(element_id: StatusLineElementID) -> impl Fn(&mut RenderContext, F)
+    where
+        F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+    {
+        match element_id {
+            helix_view::editor::StatusLineElement::Mode => Self::render_mode,
+            helix_view::editor::StatusLineElement::Spinner => Self::render_lsp_spinner,
+            helix_view::editor::StatusLineElement::FileName => Self::render_file_name,
+            helix_view::editor::StatusLineElement::FileEncoding => Self::render_file_encoding,
+            helix_view::editor::StatusLineElement::FileType => Self::render_file_type,
+            helix_view::editor::StatusLineElement::Diagnostics => Self::render_diagnostics,
+            helix_view::editor::StatusLineElement::Selections => Self::render_selections,
+            helix_view::editor::StatusLineElement::Position => Self::render_position,
+        }
+    }
+
+    fn render_mode<F>(context: &mut RenderContext, write: F)
+    where
+        F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
+    {
+        let visible = context.focused;
+
+        write(
+            context,
+            format!(
+                " {} ",
+                if visible {
+                    match context.doc.mode() {
+                        Mode::Insert => "INS",
+                        Mode::Select => "SEL",
+                        Mode::Normal => "NOR",
+                    }
+                } else {
+                    // If not focused, explicitly leave an empty space instead of returning None.
+                    "   "
+                }
+            ),
+            if visible {
+                match context.doc.mode() {
+                    Mode::Insert => Some(context.theme.get("ui.statusline.insert")),
+                    Mode::Select => Some(context.theme.get("ui.statusline.select")),
+                    Mode::Normal => Some(context.theme.get("ui.statusline.normal")),
+                }
+            } else {
+                None
+            },
+        );
+    }
```

Comment:
Should we keep all these functions within the same struct? They don't even need `Self`, would it be better if they are splitted out into functions like typed commands?

Reply:
Agree, that would also be in line with gutter components which are just functions in `gutter.rs`.

Reply:
Do you suggest completely getting rid of the `StatusLine` struct? Because none of the functions require `Self`.

I initially designed it that way assuming I would need a context in `Self`, but I figured it'd be better passing the context as a parameter.

Reply:
Anyway, since the struct wasn't really needed, I refactored it away in fae2348.